src/libjson/sample/main.c

```diff
@@ -0,0 +1,830 @@
+**
+** Copyright (c) Microsoft Corporation
+**
+** All rights reserved.
+**
+** MIT License
+**
+** Permission is hereby granted, free of charge, to any person obtaining a copy
+** of this software and associated documentation files (the ""Software""), to
+** deal in the Software without restriction, including without limitation the
+** rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+** sell copies of the Software, and to permit persons to whom the Software is
+** furnished to do so, subject to the following conditions: The above copyright
+** notice and this permission notice shall be included in all copies or
+** substantial portions of the Software.
+**
+** THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+** IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+** FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+** AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+** LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+** OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+** THE SOFTWARE.
+**
+**==============================================================================
+*/
+
+#include "structs.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <ctype.h>
+#include <json.h>
+#include <jsonprint.h>
+#include <sys/stat.h>
+#include <string.h>
+#include <stdint.h>
+#include <assert.h>
+#include <stdbool.h>
+
+#include "strings.h"
+
+const char* arg0;
+
+#define COUNTOF(ARR) (sizeof(ARR) / sizeof(ARR[0]))
+
+typedef struct json_callback_data
+{
+    header_t* hdr;
+}
+json_callback_data_t;
+
+static int _strtou64(uint64_t* x, const char* str)
+{
+    char* end;
+
+    *x = strtoull(str, &end, 10);
+
+    if (!end || *end != '\0')
+        return -1;
+
+    return 0;
+}
+
+static json_result_t _json_read_callback(
+    json_parser_t* parser,
+    json_reason_t reason,
+    json_type_t type,
+    const json_union_t* un,
+    void* callback_data)
+{
+    json_result_t result = JSON_UNEXPECTED;
+    json_callback_data_t* data = (json_callback_data_t*)callback_data;
+
+    switch (reason)
+    {
+        case JSON_REASON_NONE:
+        {
+            /* Unreachable */
+            assert(false);
+            break;
+        }
+        case JSON_REASON_NAME:
+        {
+            break;
+        }
+        case JSON_REASON_BEGIN_OBJECT:
+        {
+            break;
+        }
+        case JSON_REASON_END_OBJECT:
+        {
+            break;
+        }
+        case JSON_REASON_BEGIN_ARRAY:
+        {
+            break;
+        }
+        case JSON_REASON_END_ARRAY:
+        {
+            break;
+        }
+        case JSON_REASON_VALUE:
+        {
+            const char** path = parser->path;
+            size_t depth = parser->depth;
+            unsigned long i;
+
+            if (json_match(parser, "keyslots.#.type", &i) == JSON_OK)
```

Comment:
The code below in this file seems like a lot of boiler plate. No way to cut this down? I'm used to nlohmann's C++ library which makes it nearly fun to use JSON.

Reply:
Not that I'm aware of without much more effort. What does nlohmann's parser do to simplify this?

Reply:
I guess, the fact that it uses C++? No worries. If Christoph is happy with it, it's fine.

Reply:
Yeah, it's C++ and it follows a completely different usage pattern, i.e. the parser internals are not exposed at all during parsing, you just get a nice map of keys to values. This parser just calls the callback and we have to do all the mapping and storing ourselves, which in C is a bit cumbersome, but faster and requires less memory.

Reply:
Thank you. I see. There is a back end for the **odatalite JSON parser** that does create parse trees. We might consider bringing that part over as well at some point. It is not C++ but at least you get a fully-formed tree. I'm not sure how helpful that would be.